endaq/calc/filters.py

```diff
@@ -74,3 +102,294 @@ def butterworth(
+    :param half_order: half of the order of the filter; higher orders provide
+        more aggressive stopband reduction
+    :param tukey_percent: the alpha parameter of a preconditioning Tukey filter;
+        if 0 (default), no filter is applied
+    :param rs: tthe minimum attenuation allowed in the stopband, specified in decibels
+
+    :return: the filtered data
+
+    .. seealso::
+
+        - `SciPy Cheby1 filter design <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.cheby2.html>`_
+          Documentation for the Chebyshev type II filter design function.
+
+        - `SciPy Tukey window <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.windows.tukey.html>`_
+          Documentation for the Tukey window function used in preprocessing.
+    """
+    cutoff_freqs, filter_type = _get_filter_frequencies_type(low_cutoff, high_cutoff)
+
+    if filter_type:
+        dt = utils.sample_spacing(df)
+
+        sos_coeffs = scipy.signal.cheby2(
+            N=half_order,
+            Wn=cutoff_freqs,
+            btype=filter_type,
+            fs=1 / dt,
+            output="sos",
+            rs=rs
+        )
+        df = df.apply(functools.partial(scipy.signal.sosfiltfilt, sos_coeffs), axis=0)
+
+    if tukey_percent > 0:
+        tukey_window = scipy.signal.windows.tukey(len(df.index), alpha=tukey_percent)
+        df = df.mul(tukey_window, axis="rows")
+
+    return df           
+
+
+def ellip(
+    df: pd.DataFrame,
+    low_cutoff: Optional[float] = 1.0,
+    high_cutoff: Optional[float] = None,
+    half_order: int = 3,
+    tukey_percent: float = 0.0,
+    rp: float = 3.0,
+    rs: float = 30.0,
+) -> pd.DataFrame:
+    """
+    Apply a lowpass and/or a highpass Elliptic filter to an array.
+
+    This function uses Elliptic filter designs, and implements the filter(s)
+    as bi-directional digital biquad filters, split into second-order sections.
+
+    :param df: the input data; cutoff frequencies are relative to the
+        timestamps in `df.index`
+    :param low_cutoff: the low-frequency cutoff, if any; frequencies below this
+        value are rejected, and frequencies above this value are preserved
+    :param high_cutoff: the high-frequency cutoff, if any; frequencies above
+        this value are rejected, and frequencies below this value are preserved
+    :param half_order: half of the order of the filter; higher orders provide
+        more aggressive stopband reduction
+    :param tukey_percent: the alpha parameter of a preconditioning Tukey filter;
+        if 0 (default), no filter is applied
+    :param rp: tthe maximum ripple allowed in the passband, specified in decibels
+    :param rs: tthe minimum attenuation allowed in the stopband, specified in decibels
+
+    :return: the filtered data
+
+    .. seealso::
+
+        - `SciPy Cheby1 filter design <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.ellip.html>`_
```

Comment:
This fixes the other doctests error
```suggestion
        - `SciPy Ellip filter design <https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.ellip.html>`_
```